### PR TITLE
Add a visualization tool

### PR DESCRIPTION
**Pull request by jmcarcell**
Piggybacking on #178 (since it has been stale for a long time), I have modified
it and made the resulting graph look nicer.

The example data model in podio looks like this:

[podio.pdf](https://github.com/AIDASoft/podio/files/10747942/podio.pdf)

and EDM4hep like this:

[EDM4hep.pdf](https://github.com/AIDASoft/podio/files/10747946/gv.pdf)

BEGINRELEASENOTES
- Add a visualization tool that converts a YAML description to a graph

ENDRELEASENOTES

### CODE REVIEW

**Comment by vvolkl** [comment]
This looks great! I wonder if the graph could be made even more informative by grouping types in clusters (https://graphviz.org/Gallery/directed/cluster.html). I'm thinking of the (Monte Carlo|Raw Data| Digitization | Reco & Analysis)  categories in https://github.com/key4hep/EDM4hep/raw/master/doc/edm4hep_diagram.svg 

The info which type belongs to which cluster would have to be hardcoded in the script or added to the yaml though.

**Comment by hegner** [comment]
@jmcarcell - great that you are picking that up :-)

**Comment by tmadlener** [comment]
Thanks for reviving this.

My 2 cents:
- I think the grouping could be left as an (optional) input to the visualization tool. That way the model definition would remain unchanged, and it would even be possible to generate diagrams with different groupings. This could just be another yaml file along the lines of
```yaml
group:
  - datatype
  - otherDatatype
anotherGroup:
  - anotherType
```
- I would propose svg as output instead of pdf, because it is much easier to embed into a webpage (and one could do some interactive fancyness if desired). This should be fairly trivial to change with graphviz, I think.
- Is it possible to make the `Associations` have two arrow ends? This would make it clearer that these are not directed.
- I am not sure inverting the direction of the edges just to get the right ordering is the way to go. IMHO the diagram gets confusing in this way, because the arrows point in the opposite direction of which information is actually retrievable.
- Maybe also filtering a few datatypes from the graph would be nice to have (e.g. some of the Associations in EDM4hep)

The last two points might be addressable by a more complex grammar for the optional input yaml file. Filtering should be rather easy, by simply only using the types that are stated in the file. Defining a "layout" for the groups potentially becomes much harder. Here, I am really not sure how complex we want this tool to be, and I would rather start with a simple one that gives correct, but potentially not yet visually perfect results.

**Comment by hegner** [comment]
Thanks a lot. Looks great

**Comment by hegner** [comment]
@jmcarcell - can you fix the flake8 error of the pre-commit test?

File: tools/podio-vis [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+if __name__ == "__main__":
+  parser = argparse.ArgumentParser(description='Given a description yaml file this script generates '
+                                   'a visualization in the target directory')
+
+  parser.add_argument('description', help='yaml file describing the datamodel')
+  parser.add_argument('-d', '--dot', action='store_true', default=False,
+                      help='just write the dot file')
+  parser.add_argument('--fmt', default='svg', help='Which format to use for saving the file')
+  parser.add_argument('--filename', default='gv', help='Which filename to use for the output')
+  parser.add_argument('--extra-input', help='Extra input file for defining groups')
```
Comment (tmadlener) [inline tools/podio-vis:98]:
```suggestion
  parser.add_argument('--graph-config', help='Optional input file for configuring, e.g. the layout of the generated graph')
```
Just to be slightly more specific here on the purpose of this input

**Review by tmadlener — COMMENTED** [review]
Thanks a lot. Looks good to me as well. Just one minor comment about the naming of the `extra-input` argument.

**Review by hegner — CHANGES_REQUESTED** [review]
You would need to list graphviz as an *optional* dependency in the PODIO description
  https://github.com/AIDASoft/podio/blob/master/README.md
We want to keep the necessary dependencies of PODIO as small as possible.

Since it is optional you could even put a try/catch in the graphviz import of the new tool printing a little comment.
The really clean solution would be actually non-installing the script if graphviz isn't available. But that may be an overkill

**Comment by hegner** [comment]
@jmcarcell - thanks! The `requirements.txt` you can ignore for now.

If you make the code checker (see sys.exit) happy, I can merge :-)

**Review by hegner — APPROVED** [review]

**Pull request merged**